File: php/bin/create_tenant.php
<?php

declare(strict_types=1);

/**
 * Alta de un restaurante nuevo en la plataforma.
 *
 * Crear un tenant es una operacion de plataforma, no de tenant: no existe
 * todavia un token del cual sacar el tenant_id, y el modelo de seguridad no
 * contempla un usuario que cruce empresas (ver CLAUDE.md, principio 1). Por
 * eso vive en un script y no en un endpoint — equivalente PHP de
 * scripts/create_tenant.py.
 *
 * Uso:
 *   php bin/create_tenant.php "Pizza Napoli" \
 *       --business-type=table_service \
 *       --branch="Sede Norte" --branch-code=NOR \
 *       --admin-email=dueno@napoli.com --admin-password="una-clave-larga"
 */

use App\Core\Database;
use App\Core\Security;
use App\Domain\TenantSettings;
use App\Repositories\BranchRepository;
use App\Repositories\RoleRepository;
use App\Repositories\UserRepository;
use App\Services\TenantProvisioning;
use App\Services\TenantProvisioningError;

require dirname(__DIR__) . '/vendor/autoload.php';

$envFile = dirname(__DIR__) . '/.env';
if (is_file($envFile)) {
    foreach (file($envFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES) as $line) {
        if (str_starts_with(trim($line), '#') || !str_contains($line, '=')) {
            continue;
        }
        [$key, $value] = explode('=', $line, 2);
        $_ENV[trim($key)] = trim($value, " \t\n\r\0\x0B\"'");
    }
}

// Parseo manual en vez de getopt(): getopt() deja de escanear en el primer
// argumento que no empiece con "--", asi que con el nombre del restaurante
// primero (el uso natural: `create_tenant.php "Nombre" --branch=...`)
// ignora todas las opciones que vengan despues. Sin esa limitacion, a mano.
$opts = [];
$name = null;
foreach (array_slice($argv, 1) as $arg) {
    if (str_starts_with($arg, '--') && str_contains($arg, '=')) {
        [$key, $value] = explode('=', substr($arg, 2), 2);
        $opts[$key] = $value;
    } elseif ($name === null) {
        $name = $arg;
    }
}

$missing = array_filter(['branch', 'branch-code', 'admin-email', 'admin-password'], fn ($k) => !isset($opts[$k]));
if ($name === null || $missing !== []) {
    fwrite(STDERR, "Uso: php bin/create_tenant.php \"Nombre\" --branch=X --branch-code=Y --admin-email=Z --admin-password=W\n");
    exit(1);
}

$businessType = $opts['business-type'] ?? 'fast_food';
if (!in_array($businessType, TenantSettings::BUSINESS_TYPES, true)) {
    $valid = implode(', ', TenantSettings::BUSINESS_TYPES);
    fwrite(STDERR, "--business-type debe ser uno de: {$valid}\n");
    exit(1);
}

$adminPassword = $opts['admin-password'];
if (strlen($adminPassword) < 8) {
    fwrite(STDERR, "La clave del administrador debe tener al menos 8 caracteres\n");
    exit(1);
}

// Todo el alta en una sola transaccion: un restaurante sin usuario con el
// cual entrar no sirve de nada, asi que si algo falla no queda nada creado.
$pdo = Database::admin();
$pdo->beginTransaction();

try {
    $tenant = TenantProvisioning::createTenant(
        $name,
        $businessType,
        $opts['currency'] ?? 'COP',
    );

    $branch = (new BranchRepository($pdo))->create(
        $tenant->id,
        $opts['branch'],
        $opts['branch-code'],
        $opts['timezone'] ?? 'America/Bogota',
        null,
        null,
    );

    $adminRole = (new RoleRepository($pdo))->getByCode($tenant->id, 'admin');
    if ($adminRole === null) {
        $pdo->rollBack();
        fwrite(STDERR, "El aprovisionamiento no dejo un rol admin; se aborta\n");
        exit(1);
    }

    (new UserRepository($pdo, new RoleRepository($pdo)))->create(
        $tenant->id,
        $adminRole->id,
        $branch->id,
        $opts['admin-name'] ?? 'Administrador',
        $opts['admin-email'],
        Security::hashPassword($adminPassword),
    );
    $pdo->commit();
} catch (TenantProvisioningError|\Throwable $e) {
    $pdo->rollBack();
    fwrite(STDERR, "No se pudo crear el restaurante: {$e->getMessage()}\n");
    exit(1);
}

echo "Restaurante '{$name}' creado\n";
echo "  tenant_id : {$tenant->id}\n";
echo "  sucursal  : {$branch->name} ({$branch->code})\n";
echo "  admin     : {$opts['admin-email']}\n";

File: scripts/create_tenant.py
"""Alta de un restaurante nuevo en la plataforma.

Crear un tenant es una operacion de plataforma, no de tenant: no existe
todavia un token del cual sacar el tenant_id, y el modelo de seguridad no
contempla un usuario que cruce empresas (ver CLAUDE.md, principio 1). Por
eso vive en un script y no en un endpoint.

Deja el restaurante listo para operar: estados segun su modelo de negocio,
rol admin, impuesto por defecto, primera sucursal y primer usuario.

    python scripts/create_tenant.py "Pizza Napoli" \
        --business-type table_service \
        --branch "Sede Norte" --branch-code NOR \
        --admin-email dueno@napoli.com --admin-password "una-clave-larga"
"""

import argparse
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

# Conexion administrativa: crear una empresa es previo a que exista tenant
# alguno, asi que no puede pasar por el rol restringido que sirve requests.
from app.core.database import AdminSessionLocal  # noqa: E402
from app.core.security import hash_password  # noqa: E402
from app.domain.tenant_settings import DEFAULTS_BY_BUSINESS_TYPE  # noqa: E402
from app.repositories import branch_repository, role_repository, user_repository  # noqa: E402
from app.services.tenant_provisioning import create_tenant  # noqa: E402


def main() -> int:
    parser = argparse.ArgumentParser(description="Da de alta un restaurante nuevo")
    parser.add_argument("name", help="Nombre del restaurante")
    parser.add_argument(
        "--business-type",
        default="fast_food",
        choices=sorted(DEFAULTS_BY_BUSINESS_TYPE),
        help="Define el flujo de estados y la configuracion inicial",
    )
    parser.add_argument("--currency", default="COP")
    parser.add_argument("--branch", required=True, help="Nombre de la primera sucursal")
    parser.add_argument("--branch-code", required=True, help="Codigo corto, prefijo del numero de pedido")
    parser.add_argument("--timezone", default="America/Bogota")
    parser.add_argument("--admin-name", default="Administrador")
    parser.add_argument("--admin-email", required=True)
    parser.add_argument("--admin-password", required=True)
    args = parser.parse_args()

    if len(args.admin_password) < 8:
        print("La clave del administrador debe tener al menos 8 caracteres", file=sys.stderr)
        return 1

    db = AdminSessionLocal()
    try:
        tenant = create_tenant(
            db, name=args.name, business_type=args.business_type, currency=args.currency
        )

        branch = branch_repository.create(
            db,
            tenant_id=tenant.id,
            name=args.branch,
            code=args.branch_code,
            timezone=args.timezone,
            address=None,
            phone=None,
        )

        admin_role = role_repository.get_by_code(db, tenant.id, "admin")
        if admin_role is None:
            print("El aprovisionamiento no dejo un rol admin; se aborta", file=sys.stderr)
            db.rollback()
            return 1

        user_repository.create(
            db,
            tenant_id=tenant.id,
            role_id=admin_role.id,
            branch_id=branch.id,
            name=args.admin_name,
            email=args.admin_email,
            password_hash=hash_password(args.admin_password),
        )
        db.commit()

        # commit() expira los atributos: hay que leerlos antes de cerrar la
        # sesion o el refresh perezoso reventaria contra una sesion muerta.
        summary = (str(tenant.id), branch.name, branch.code)
    except Exception as exc:
        db.rollback()
        print(f"No se pudo crear el restaurante: {exc}", file=sys.stderr)
        return 1
    finally:
        db.close()

    tenant_id, branch_name, branch_code = summary
    print(f"Restaurante '{args.name}' creado")
    print(f"  tenant_id : {tenant_id}")
    print(f"  sucursal  : {branch_name} ({branch_code})")
    print(f"  admin     : {args.admin_email}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
